pkg/clustertree/knode-manager/controllers/secret.go

```diff
@@ -0,0 +1,291 @@
+	}
+	klog.Info("kosmos sync master secret cache success")
+
+	if !cache.WaitForCacheSync(stopCh, s.clientSynced) {
+		return fmt.Errorf("kosmos cannot sync client secret cache")
+	}
+	klog.Info("kosmos sync client secret cache success")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(s.syncClientSecret, 0, stopCh)
+		go wait.Until(s.resyncClientSecret, 0, stopCh)
+	}
+
+	<-stopCh
+	return nil
+}
+
+// sync master secret -> client secret
+func (s *SecretController) syncClientSecret() {
+	secret2key, result := s.masterSecretQueue.Get()
+	if result {
+		return
+	}
+	defer s.masterSecretQueue.Done(secret2key)
+	ns, secretName, err := cache.SplitMetaNamespaceKey(secret2key.(string))
+	if err != nil {
+		s.masterSecretQueue.Forget(secret2key.(string))
+		return
+	}
+
+	klog.Infof("kosmos sync secret: start sync secret, secret: %v", secret2key.(string))
+	secretLifecycle := false
+
+	secretInstance, err := s.masterLister.Secrets(ns).Get(secretName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("kosmos sync: get secret from master list failed, error: %v", err)
+			return
+		}
+		secretLifecycle = true
+	}
+	defer func() {
+		if err != nil {
+			s.masterSecretQueue.AddRateLimited(secret2key)
+			return
+		}
+		s.masterSecretQueue.Forget(secret2key)
+	}()
+	if secretLifecycle || secretInstance.DeletionTimestamp != nil {
+		err = s.handler.DeleteClientSecret(ns, secretName)
+		if err != nil {
+			klog.Errorf("kosmos sync: delete secret from client list failed, secret: %v", secretName)
+			return
+		}
+		klog.Infof("kosmos sync: delete secret from client list succeed, secret: %v", secretName)
+		return
+	}
+
+	klog.Infof("Secret %v/%v to be update", ns, secretName)
+	s.syncSecretHandler(secretInstance)
+}
+
+// resync master secret -> client secret (Prevents the client secret from being incorrectly modified)
+func (s *SecretController) resyncClientSecret() {
+	secret2key, result := s.clientSecretQueue.Get()
+	if result {
+		return
+	}
+	defer s.clientSecretQueue.Done(secret2key)
+	ns, secretName, err := cache.SplitMetaNamespaceKey(secret2key.(string))
+	if err != nil {
+		s.clientSecretQueue.Forget(secret2key.(string))
+		return
+	}
+
+	var masterSecret *corev1.Secret
+	var clientSecret *corev1.Secret
+
+	clientSecret, err = s.clientLister.Secrets(ns).Get(secretName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("kosmos sync: get secret from client list failed, error: %v", err)
+			return
+		}
+		masterSecretLifecycle := false
+		masterSecret, err = s.masterLister.Secrets(ns).Get(secretName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				klog.Errorf("kosmos sync: get secret from master list failed, error: %v", err)
+				return
+			}
+			masterSecretLifecycle = true
+		}
+		// The client deletes its own secret. We need to synchronize the master's secret to the client
+		if !masterSecretLifecycle && masterSecret.DeletionTimestamp == nil {
```

Comment:
what's meaning of masterSecretLifecycle?

Reply:
masterSecretLifecycle means whether the secret exists in the master. (deleted or not deleted)

Reply:
masterSecretLifecycle should be false if the secret has been deleted

Reply:
I got it. And i haved fixed it